emeis/core/permissions.py

```diff
@@ -0,0 +1,113 @@
+        fn._permissions.append(model)
+        return fn
+
+    return decorate
+
+
+def object_permission_for(model):
+    """Decorate function to overwriting object permission of specific model."""
+
+    def decorate(fn):
+        if not hasattr(fn, "_object_permissions"):
+            fn._object_permissions = []
+        fn._object_permissions.append(model)
+        return fn
+
+    return decorate
+
+
+class BasePermission(object):
+    """Basic permission class to be extended by any permission implementation.
+
+    In combination with the decorators `@permission_for` and `@object_permission_for` a custom
+    permission class can define permission on basis of models.
+
+    Per default it returns True but default can be overwritten to define a permission for
+    `BaseModel`
+
+    A custom permission class could look like this:
+    ```
+    >>> from emeis.core.permissions import BasePermission
+    ... from emeis.core.models import BaseModel, User, Scope
+    ...
+    ... class CustomPermission(BasePermission):
+    ...     @permission_for(BaseModel)
+    ...     def has_permission_default(self, request):
+    ...         # change default permission to False when no more specific
+    ...         # permission is defined.
+    ...         return False
+    ...
+    ...     @permission_for(Scope)
+    ...     def has_permission_for_save_form(self, request):
+    ...         return True
+    ...
+    ...     @object_permission_for(User)
+    ...     def has_object_permission_for_save_form(self, request, instance):
+    ...         return request.user.username == 'admin'
+    """
+
+    def __init__(self):
+        perm_fns = inspect.getmembers(self, lambda m: hasattr(m, "_permissions"))
+        perm_muts = [
+            permission.__name__ for _, fn in perm_fns for permission in fn._permissions
+        ]
+        perm_muts_dups = list_duplicates(perm_muts)
+        if perm_muts_dups:
+            raise ImproperlyConfigured(
+                f"`permission_for` defined multiple times for "
+                f"{', '.join(perm_muts_dups)} in {str(self)}"
+            )
+        self._permissions = {
+            permission: fn for _, fn in perm_fns for permission in fn._permissions
+        }
+
+        obj_perm_fns = inspect.getmembers(
+            self, lambda m: hasattr(m, "_object_permissions")
+        )
+        obj_perm_muts = [
+            permission.__name__
+            for _, fn in obj_perm_fns
+            for permission in fn._object_permissions
+        ]
+        obj_perm_muts_dups = list_duplicates(obj_perm_muts)
+        if obj_perm_muts_dups:
+            raise ImproperlyConfigured(
+                f"`object_permission_for` defined multiple times for "
+                f"{', '.join(obj_perm_muts_dups)} in {str(self)}"
+            )
+        self._object_permissions = {
+            permission: fn
+            for _, fn in obj_perm_fns
+            for permission in fn._object_permissions
+        }
```

Comment:
Not important: I think the above code could be extracted, its twice the same pattern. But it might decrease readablity.

Reply:
For the sake of readability, I'd prefer to leave it as is.